### PR TITLE
Add ch.babs.kulturgueter-anhoerung

### PR DESCRIPTION
**Pull request by b4l**
Test link: https://mf-chsdi3.dev.bgdi.ch/ltbal_kulturgueter/shorten/8c2edcb1f7

WMS: https://github.com/geoadmin/wms-mapfile_include/pull/565
Script: https://github.com/geoadmin/bgdi-scripts/pull/731

### CODE REVIEW

File: chsdi/templates/htmlpopup/kgs_anhoerung.mako [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+            </td>
+	    </tr>
+    %endif
+    % if c['attributes']['link_uri'] is not None:
+        <tr>
+          <th class="cell-left">${_('legalregulationlink')}</th>
+            <td><a href="${c['attributes']['link_uri']}">${c['attributes']['link_title']}</a></td>
+        </tr>
+    % endif
+    % if c['attributes']['link_2_uri'] is not None:
```
Comment (marionb) [inline chsdi/templates/htmlpopup/kgs_anhoerung.mako:103]: looks like these if clause do not work properly the rows `weitere informationen` appeares 3 times out of which two are empty.
https://mf-chsdi3.dev.bgdi.ch/ltbal_kulturgueter/rest/services/ech/MapServer/ch.babs.kulturgueter-anhoerung/3920/extendedHtmlPopup?lang=de

File: chsdi/templates/htmlpopup/kgs_anhoerung.mako [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+                    $('#blueimp-gallery-title').html(($($('.thumbnail-container').children('.thumbnail')[index]).children('div').html()));
+                }},
+                links = this.getElementsByTagName('a');
+              blueimp.Gallery(links, options);
+            });
+        });
+    </script>
+
+
+    <table class="table-with-border kernkraftwerke-extended">
```
Comment (marionb) [inline chsdi/templates/htmlpopup/kgs_anhoerung.mako:54]: `kernkraftwerke-extended` This is a funny class name. Is it possible to generalize this name if it is needed here?

File: chsdi/templates/htmlpopup/kgs_anhoerung.mako [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+                <div class="thumbnail">
+                    <a href="https://${dataGeoAdminHost}/${dataPath}/image/kgs_${pic[0]}_${pic[1]}.jpg">
+                        <img class="image" src="https://${dataGeoAdminHost}/${dataPath}/image/kgs_${pic[0]}_${pic[1]}.jpg" />
+                    </a>
+                    <div>${pic[3] or ''} - ${pic[2] or ''}</div>
+                </div>
+            %endfor
+            </div>
+        </div>
+    <table class="kernkraftwerke-extended">
```
Comment (marionb) [inline chsdi/templates/htmlpopup/kgs_anhoerung.mako:137]: samme comment as in line 54

**Review by marionb — CHANGES_REQUESTED** [review]
Othere than what I added as comments in the code this is working fine and looks good to me.

**Comment by marionb** [comment]
One thing to note:
I do not know if you have control over it and if it is already suposed to work but the link to the object in the tooltip is not working.

**Comment by b4l** [comment]
> One thing to note:
> I do not know if you have control over it and if it is already suposed to work but the link to the object in the tooltip is not working.

Since the link is pointing to prod, this is expected to fail.

File: chsdi/templates/htmlpopup/kgs_anhoerung.mako (hunk 1 above)
Comment (b4l) [reply chsdi/templates/htmlpopup/kgs_anhoerung.mako:103]: This is due to the data delivering an empty string. It is the same for ch.babs.kulturgueter (babs.mako & vbs_master.babs.kgs) which has just been copied. The latter is as well the origin of the other remarks and fancy naming. @mariokeusen is this ok or would it be desirable to adjust this?

Comment (marionb) [reply chsdi/templates/htmlpopup/kgs_anhoerung.mako:103]: OK I see, you could check for an empty string as well. But it is probably better to check what is really requested. It just looks a bit strange to me from a user perspective.

File: chsdi/templates/htmlpopup/kgs_anhoerung.mako [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+        csv_file = None
+        try:
+            csv_file = urlopen(csv_url)
+            reader = csv.reader(csv_file, delimiter =';')  # creates the reader object
+            pic_list = []
+            for i, row in enumerate(reader):   # iterates the rows of the file in orders
+                if i == 0: # The first row is NUMMER;BILDNR;FOTOGRAF;COPYRIGHT and cannot be parsed.
+                    continue
+                if int(row[0]) == c['featureId']:
+                    pic_list.append(map(lambda x: x.decode('utf-8'), row))
```
Comment (ltclm) [inline chsdi/templates/htmlpopup/kgs_anhoerung.mako:32]: maybe we should switch to utf in kgs.mako too.

File: chsdi/templates/htmlpopup/kgs_anhoerung.mako [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+<%def name="extended_info(c, lang)">
+    <%
+        c['stable_id'] = True
+        objarts = c['attributes']['objektart'].split(',')
+        import csv
+        from urllib2 import urlopen
+        dataGeoAdminHost = request.registry.settings['datageoadminhost']
+        dataPath = 'ch.babs.kulturgueter'
+        csv_url = "https://" + dataGeoAdminHost + "/ch.babs.kulturgueter-anhoerung/image/meta.txt"
```
Comment (ltclm) [inline chsdi/templates/htmlpopup/kgs_anhoerung.mako:22]: is it really planned to use the same meta.txt as for the official layer?
https://data.geo.admin.ch/ch.babs.kulturgueter-anhoerung/image/meta.txt

**Review by ltclm — CHANGES_REQUESTED** [review]
looks good to me, just two questions concerning the relationship with the other layer.

if you have time you could address the following points:
there is a lot of redundant stuff in the model and the mako.

- **model**
you could create a parent class with the identical properties and inherit this class to the two child classes.

- **mako**
you could do the same here, example: cadastralwebmap.mako

File: chsdi/templates/htmlpopup/kgs_anhoerung.mako (hunk 5 above)
Comment (b4l) [reply chsdi/templates/htmlpopup/kgs_anhoerung.mako:22]: No its a new one.

File: chsdi/templates/htmlpopup/kgs_anhoerung.mako (hunk 4 above)
Comment (ltclm) [reply chsdi/templates/htmlpopup/kgs_anhoerung.mako:32]: ok, its not the same file so this comment can be ignored.

File: chsdi/templates/htmlpopup/kgs_anhoerung.mako (hunk 5 above)
Comment (ltclm) [reply chsdi/templates/htmlpopup/kgs_anhoerung.mako:22]: ok.

File: chsdi/templates/htmlpopup/kgs_anhoerung.mako (hunk 2 above)
Comment (b4l) [reply chsdi/templates/htmlpopup/kgs_anhoerung.mako]: Not changing this in this pull request as it touches around 30 makos.

File: chsdi/templates/htmlpopup/kgs_anhoerung.mako (hunk 3 above)
Comment (b4l) [reply chsdi/templates/htmlpopup/kgs_anhoerung.mako]: Not changing this in this pull request as it touches around 30 makos.

**Review by marionb — APPROVED** [review]
LGTM now. Thanks!

**Review by faselm — COMMENTED** [review]
@b4l Not sure if it's normal but I cannot see the symbology with the test link any more (data, legends, tooltip look good)...

**Comment by b4l** [comment]
> @b4l Not sure if it's normal but I cannot see the symbology with the test link any more (data, legends, tooltip look good)...

It's because my personal wms instance is not serving the layer any more.

**Comment by rebert** [comment]
@b4l it would be great to have a working testlink. There is the possibility to configure a wms 4 abnahme like dev_ltrea_abnahme

**Comment by b4l** [comment]
@faselm @rebert the link is working again.

**Comment by rebert** [comment]
of course it is up to the Datenherr. Me personally, I would have put x for easting and y for nothing...

**Review by rebert — APPROVED** [review]

**Review by ltclm — COMMENTED** [review]
I know its coming from the data delivery but the coordinates of the official layer are in lv95 and the coordinates of the revision layer are in lv03:
image

File: chsdi/templates/htmlpopup/kgs.mako [hunk 6]
```diff
@@ -15,21 +15,18 @@
     <%
         c['stable_id'] = True
         objarts = c['attributes']['objektart'].split(',')
+        gemeinde_ehemalig = c['attributes']['gemeinde_ehemalig'] if c['attributes']['gemeinde_ehemalig'].startswith("(") else "("+c['attributes']['gemeinde_ehemalig']+")"
```
Comment (ltclm) [inline chsdi/templates/htmlpopup/kgs.mako:18]: sure about that :-)

**Review by ltclm — APPROVED** [review]
lgtm, thank you.
its by design that the two layers are using the same image and pdf source on s3?
* ch.babs.kulturgueter-anhoerung -> https://data.geo.admin.ch/ch.babs.kulturgueter/PDF, https://data.geo.admin.ch/ch.babs.kulturgueter/images
* ch.babs.kulturgueter -> https://data.geo.admin.ch/ch.babs.kulturgueter/PDF, https://data.geo.admin.ch/ch.babs.kulturgueter/images

File: chsdi/templates/htmlpopup/kgs.mako (hunk 6 above)
Comment (b4l) [reply chsdi/templates/htmlpopup/kgs.mako:18]: ja

**Comment by b4l** [comment]
> its by design that the two layers are using the same image and pdf source on s3?

yes, the mapping is different though (from the data and the excel)

**Pull request merged**